compiler/rustc_parse/src/lexer/unescape_error_reporting.rs

```diff
@@ -55,11 +55,23 @@ pub(crate) fn emit_unescape_error(
                 ("", "if you meant to write a `str` literal, use double quotes")
             };
 
-            handler
-                .struct_span_err(
+            let mut handler = handler.struct_span_err(
+                span_with_quotes,
+                "character literal may only contain one codepoint",
+            );
+            if lit.chars().count() == 2
+                && ('\u{FE00}'..='\u{FE0F}').contains(&lit.chars().last().unwrap())
```

Comment:
Don't forget [Variation Selectors Supplement](https://en.wikipedia.org/wiki/Variation_Selectors_Supplement) and the [Mongolian Variation Selectors](https://en.wikipedia.org/wiki/Mongolian_(Unicode_block)#Variations_and_vowel_separation)... indeed, perhaps [`unicode_normalization::char::is_combining_mark`](https://docs.rs/unicode-normalization/0.1.19/unicode_normalization/char/fn.is_combining_mark.html) could offer a more general approach?

Reply:
I have expanded my implementation now to print a note for all combining marks, as you suggested. I've also added a suggestion to use the normalized form if the character can be represented as one codepoint (as in [this](https://www.unicode.org/reports/tr15/images/UAX15-NormFig4.jpg) example).

Reply:
Very nice!